_posts/2020-05-27-mlr3keras-boston-housing-task/mlr3keras-boston-housing-task.Rmd

```diff
@@ -0,0 +1,332 @@
+* `cb_tb`: Tensorboard callback
+* `cb_lr_log`: Learning rate logger callback
+* `LogMetrics`: Batch-wise Metrics Logger Callback
+* `SetLogLR`: Batch-wise Metrics Setter Callback
+
+Initialize early stopping like this:
+
+```{r mlr3cb}
+# early_stop <- callback_early_stopping(monitor = "val_loss", patience = 20) # keras syntax
+early_stop <- cb_es(monitor = "val_loss", patience = 20) # mlr3 syntax
+```
+
+
+### Set Up the Learner
+
+We initialize our learner.
+Note, that we clone the `R6` object `keras_500` learner to a second object `keras_es` that we will use later on.
+
+```{r po}
+keras_500 = GraphLearner$new(po_scale %>>% learner)
+keras_es = keras_500$clone(deep = TRUE)
+```
+
+# Training
+
+Before we train the learner we divide observations randomly into train and test set.
+The training set contains 404 observations and the test set 102.
+Now we can train our learner, using the `training_id`.
+Reminder: The `callback` setting vizualizes training progress by a dot after each `epoch`.
+
+```{r trainingI}
+train_id <- sample(boston_task$row_ids, 404L)
+test_id <- setdiff(boston_task$row_ids, train_id)
+keras_500$train(boston_task, row_ids = train_id)
+```
+
+### Visualize training progress
+
+Lets' plot each `epochs`' saved history.
+The plot shows the `MAE` on training and validation set for each epoch.
+
+```{r plotI}
+history = keras_500$model$regr.keras$model$history
+plot(history, metrics = "mean_absolute_error", smooth = FALSE) +
+  coord_cartesian(ylim = c(0, 5))
+```
+
+The validation curve noticeably flattens after about 200 epochs.
+So `epochs = 500L` seems to be set to high.
+In this case we can implement the early stopping callback introduced in the previous chapter.
+`early_stop` will stop the model if it shows no improvement in `val_loss` within
+20 `epochs`.
+
+### Train with early stopping
+
+We train exactly the same keras model, using early stopping `callback`.
+
+```{r earlystopping}
+keras_es$param_set$values$regr.keras.callbacks = list(early_stop, print_dot_callback)
+keras_es$param_set$values$regr.keras.model = build_model()
+keras_es$train(boston_task, row_ids = train_id)
+history = keras_es$model$regr.keras$model$history
+plot(history, metrics = "mean_absolute_error", smooth = FALSE) +
+  coord_cartesian(xlim = c(0, 500), ylim = c(0, 5))
+```
+
+Note that each `epoch` until stopping is exactly the same as in `keras_500`.
+
+# Predict and evaluate
+
+In the last step we predict on the test set using `keras_es`.
+Prediction is as easy as with any other learner.
+For prediction we take the `test_id`.
+
+```{r predict}
+predict_boston = keras_es$predict(boston_task, row_ids = test_id)
+head(predict_boston$data$tab, n = 3)
+predict_boston$score(msr("regr.mae"))
+```
+
+```{r pastemae, echo=FALSE}
+paste0(
+  "The models average error is at ",
+  round(predict_boston$score(msr("regr.mae")) * 1000, 2),
+  "$")
+```
+
+
+# Set Seed
+
+`Keras R Studio` is a system running code on multiple platforms.
```

Comment:
I don't like this sentence. Instead:
`Keras` can run code across multiple devices.

Reply:
Write one sentence about sources of stochasticity: 
- initial weights 
- ordering in batches
- stochasticity in model architecture
- parallelism
True eproducibility can only be guaranteed if code is not run in parallel and initial weights are fixed.